### DISCUSSION TITLE
"Error while executing JS" shows up Intermittently instead of data

### DISCUSSION THREAD

**Discussion by stellarfinn** [Budibase/budibase#9716] [Help]
Hi everyone, I have some JS manipulating data coming in before its displayed. 
About 25% of the time I get a failure error "Error while executing JS". (See attached screenshot). 

There doesn't appear to be any JS code issues as it mostly works without issue, but the intermittent error definitely isn't feesible for everyday use. 

Can anyone provide some insight into why this is happening?
![Screenshot 2023-02-16 093836](https://user-images.githubusercontent.com/99847728/219430404-c262cdf6-dcbb-4270-b3e9-fa8d588d7c05.jpg)

**Comment by melohagan** [comment]
Hey @stellarfinn 

Could you share your binding code and a screenshot of your component tree please?

**Reply by stellarfinn** [reply]
The component tree is a bit long so I attached two screenshots. The binding code (JS) is below for the "PER UNIT" column. 

`var wholesale = $("New Repeater block.Products.Price_Wholesale");
var units = $("New Repeater block.Products.CasePack");
var msrp = $("New Repeater block.Products.Price_MSRP");
var discount = $("New Form.Fields.discount");
var result = 0;
var case_only = $("New Repeater block.Products.case_only");

if (wholesale > 0){
	var price = wholesale;
} else {
	price = msrp;
}

if (units == 1) {
	return "—";
}

if (case_only == true && units > 1) {
	discount = 1-(discount / 100);
	result = price * discount;
	result = result / units;
	result = result.toFixed(2);
	return "$" + result;
}

discount = 1-(discount / 100);
result = price * discount;
result = result.toFixed(2);
return "$" + result;`

![component 2](https://user-images.githubusercontent.com/99847728/219460784-6a59e961-3180-47c3-b61f-d4cf75685921.jpg)
![component 1](https://user-images.githubusercontent.com/99847728/219460789-bd2924f3-bf5c-43f2-b161-a2eadf2f794e.jpg)

**Reply by stellarfinn** [reply]
Also, here's an updated view of the screen. I completed the product information so there were not any missing fields. 

![screencapture-ciaoimports-budibase-app-app-portal-2023-02-16-12_00_06](https://user-images.githubusercontent.com/99847728/219461982-28186ede-6d25-4666-ab85-583fb9e27af0.png)

**Reply by melohagan** [reply]
The only thing I can think of off the top of my head is this line:
`result = result.toFixed(2);`

If the result happened to be null/undefined then the JS would error. Might be worth trying this:
`result = result?.toFixed(2);`

**Comment by stellarfinn** [comment]
I cleaned up the JS and re-addressed the logic to handle the different
scenarios for pricing. Then I moved the JS from the block value and into
the conditional section. I’ll share this once I get to my computer. While
it seems to have stopped the JS errors, I have confirmed that having JS is
slowing down the page load randomly and/or significantly.

Follow up shortly!

On Fri, Feb 17, 2023 at 3:32 AM melohagan ***@***.***> wrote:

> The only thing I can think of off the top of my head is this line:
> result = result.toFixed(2);
>
> If the result happened to be null/undefined then the JS would error. Might
> be worth trying this:
> result = result?.toFixed(2);
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Budibase/budibase/discussions/9716#discussioncomment-5008547>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXZY4MEQ2KFWZQHOCSFJWVLWX5HTHANCNFSM6AAAAAAU6M4UW4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Sincerely,

*Bobby Segurola*
*(954) 736-7836*
*--*

P Please consider the environment before printing this email.

CONFIDENTIALITY NOTICE: This e-mail transmission, and any documents,
files or previous e-mail messages attached to it, may contain confidential
information that is legally privileged. If you are not the intended
recipient, or a person responsible for delivering it to the intended
recipient, you are hereby notified that any disclosure, copying,
distribution or use of any of the information contained in or attached
to this message is STRICTLY PROHIBITED.

**Comment by stellarfinn** [comment]
So what I've done is split up and move the JS from the Text Field... (screenshot below)... to the Conditions Field..

![text-field](https://user-images.githubusercontent.com/99847728/219787944-6d16bfda-ef86-414f-b27a-e2f51ccbb8db.jpg)

Instead of having a longer, complex JS script to parse the data depending on different scenarios, I created Conditions and then put pieces of the original JS into each so that only specific JS is run if the Condition is valid. For what. ever reason, I no longer get "Error while executing JS" errors doing it this way. I'm going to try to improve on this where possible by converting the JS into Handlebar logic and not using JS at all. I'm not sure how to approach that since there's still logic in the Conditions JS that I need to figure out how to do. 

Doing it this way occasionally results in faster load times and in other times not so I can't confirm if this is also a solution for load time issues. 

![conditions](https://user-images.githubusercontent.com/99847728/219788158-be03a728-2da8-4a5f-9262-d2e4237ebd50.jpg)
